src/DDTrace/OpenTelemetry/SpanBuilder.php

```diff
@@ -0,0 +1,219 @@
+    {
+        if ($timestampNanos >= 0) {
+            $this->startEpochNanos = $timestampNanos / 1000000000;
+        }
+
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function setSpanKind(int $spanKind): SpanBuilderInterface
+    {
+        $this->spanKind = $spanKind;
+
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function startSpan(): SpanInterface
+    {
+        $this->applySpanKind();
+
+        $parentContext = Context::resolve($this->parentContext);
+        $parentSpan = Span::fromContext($parentContext);
+        $parentSpanContext = $parentSpan->getContext();
+
+        $span = $parentSpanContext->isValid() ? null : \DDTrace\start_trace_span($this->startEpochNanos);
+        $traceId = $parentSpanContext->isValid() ? $parentSpanContext->getTraceId() : \DDTrace\root_span()->traceId;
+
+        $samplingResult = $this
+            ->tracerSharedState
+            ->getSampler()
+            ->shouldSample(
+                $parentContext,
+                $traceId,
+                $this->spanName,
+                $this->spanKind,
+                $this->attributesBuilder->build(),
+                $this->links,
+            );
+
+        $span = $span ?? \DDTrace\start_trace_span($this->startEpochNanos);
+        $samplingDecision = $samplingResult->getDecision();
+        $sampled = SamplingResult::RECORD_AND_SAMPLE === $samplingDecision;
+        $samplingResultTraceState = $samplingResult->getTraceState();
+
+        if ($parentSpanContext->isValid()) {
+            // Traceparent: {2:version}-{32:hex trace id}-{16:hex parent id}-{2:trace_flags}, version always being '00'
+            // Since parentSpanContext is valid, the trace identifiers are guaranteed to be in hexadecimal format
+            $parentId = $parentSpanContext->getSpanId();
+            $traceFlags = $sampled ? '01' : '00';
+            $traceParent = "00-$traceId-$parentId-$traceFlags";
+            \DDTrace\consume_distributed_tracing_headers([
+                'traceparent' => $traceParent,
+                'tracestate' => (string) $samplingResultTraceState, // __toString() is implemented in TraceState
+            ]);
+        } elseif ($samplingResultTraceState) {
+            $samplingResultTraceState = $samplingResultTraceState->without('dd');
+            \DDTrace\root_span()->tracestate = (string) $samplingResultTraceState;
+        }
+
+        $hexSpanId = $span->hexId();
+        $spanContext = DDTraceAPI\SpanContext::createFromLocalSpan($span, $sampled, $traceId, $hexSpanId);
+
+        if (!in_array($samplingDecision, [SamplingResult::RECORD_AND_SAMPLE, SamplingResult::RECORD_ONLY], true)) {
+            return Span::wrap($spanContext);
+        }
+
+        $span->resource = $this->spanName; // OTel.name => DD.resource
+
+        $attributesBuilder = clone $this->attributesBuilder; // According to OTel's spec, attributes can't be changed after span creation...
```

Comment:
Note to self: cloning may not be necessary anymore

Reply:
You're still assigning sampling attributes, thereby having a side-effect... I'd leave it as is.